### PR TITLE
Add back to top button on beatmap listing

### PR DESCRIPTION
**Pull request by notbakaneko**
partly addresses #3370

Sticky search bar is going to be done separately since I expect implementing that to be much messier than it looks.

- should only scroll if the current scroll position is below the top of search panel.
- clicking the button should preserve the previous scroll position and clicking on it again should restore the scroll position.
- scrolling after clicking "back to top" will clear the saved scroll position.
- new search or changing the sort order should cause a reset.

---

Allow users to scroll the page back up to the search input on the beatmap listing page by clicking a button.

### CODE REVIEW

File: resources/assets/lib/back-to-top.coffee [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      @setState lastScrollY: null
+    else
+      scrollY = if @props.anchor? then $(@props.anchor.current).offset().top else 0
+      if window.scrollY > scrollY
+        @setState lastScrollY: window.scrollY
+
+        @destroyTooltip()
+        window.scrollTo(window.scrollX, scrollY)
+
+        Timeout.set 0, () =>
```
Comment (nanaya) [inline resources/assets/lib/back-to-top.coffee]: sometimes triggered when scrolling to top/anchor on firefox :confused:

File: resources/assets/lib/back-to-top.coffee [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+  onClick: (_e) =>
+    if @state.lastScrollY?
+      @destroyTooltip()
+      window.scrollTo(window.scrollX, @state.lastScrollY)
+
+      @setState lastScrollY: null
+    else
+      scrollY = if @props.anchor? then $(@props.anchor.current).offset().top else 0
+      if window.scrollY > scrollY
```
Comment (nanaya) [inline resources/assets/lib/back-to-top.coffee]: `window.pageYOffset` for better compatibility.

**Comment by nanaya** [comment]
Just remembered about this but do you think the button should be hidden when around the top and no saved position?

File: resources/assets/lib/back-to-top.coffee [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    document.removeEventListener 'scroll', @onScroll
+
+
+  render: =>
+    button
+      className: 'back-to-top'
+      'data-tooltip-float': 'fixed'
+      onClick: @onClick
+      title: if @state.lastScrollY? then osu.trans('common.buttons.back_to_previous') else osu.trans('common.buttons.back_to_top')
+      ref: @container
```
Comment (nanaya) [inline resources/assets/lib/back-to-top.coffee]: not used anymore

**Comment by notbakaneko** [comment]
Now with stupid workaround for firefox

**Comment by notbakaneko** [comment]
I thought about hiding the button if the viewport is above the anchor point but couldn't be bothered adding more things to watch for its position, although now that it's using an `IntersectionObserver`, it'd be simpler to handle

**Pull request merged**